client/pluginmanager/csimanager/instance.go

```diff
@@ -0,0 +1,209 @@
+		fingerprintController: p.Type == dynamicplugins.PluginTypeCSIController,
+
+		shutdownCtx:         ctx,
+		shutdownCtxCancelFn: cancelFn,
+		shutdownCh:          make(chan struct{}),
+	}
+}
+
+func (i *instanceManager) run() {
+	c, err := csi.NewClient(i.info.ConnectionInfo.SocketPath)
+	if err != nil {
+		i.logger.Error("failed to setup instance manager client", "error", err)
+		close(i.shutdownCh)
+		return
+	}
+	i.client = c
+
+	go i.runLoop()
+}
+
+func (i *instanceManager) requestCtxWithTimeout(timeout time.Duration) (context.Context, context.CancelFunc) {
+	return context.WithTimeout(i.shutdownCtx, timeout)
+}
+
+func (i *instanceManager) runLoop() {
+	// basicInfo holds a cache of data that should not change within a CSI plugin.
+	// This allows us to minimize the number of requests we make to plugins on each
+	// run of the fingerprinter, and reduces the chances of performing overly
+	// expensive actions repeatedly, and improves stability of data through
+	// transient failures.
+	var basicInfo *structs.CSIInfo
+
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-i.shutdownCtx.Done():
+			if i.client != nil {
+				i.client.Close()
+				i.client = nil
+			}
+			close(i.shutdownCh)
+			return
+		case <-timer.C:
+			ctx, cancelFn := i.requestCtxWithTimeout(managerFingerprintInterval)
+
+			if basicInfo == nil {
+				info, err := i.buildBasicFingerprint(ctx)
+				if err != nil {
+					// If we receive a fingerprinting error, update the stats with as much
+					// info as possible and wait for the next fingerprint interval.
+					info.HealthDescription = fmt.Sprintf("failed initial fingerprint with err: %v", err)
+					cancelFn()
+					i.updater(i.info.Name, basicInfo)
+					timer.Reset(managerFingerprintInterval)
+					continue
+				}
+
+				// If fingerprinting succeeded, we don't need to repopulate the basic
+				// info and we can stop here.
+				basicInfo = info
+			}
+
+			info := basicInfo.Copy()
+			var fp *structs.CSIInfo
+			var err error
+
+			if i.fingerprintNode {
+				fp, err = i.buildNodeFingerprint(ctx, info)
+			} else if i.fingerprintController {
+				fp, err = i.buildControllerFingerprint(ctx, info)
+			}
+
+			if err != nil {
+				info.Healthy = false
+				info.HealthDescription = fmt.Sprintf("failed fingerprinting with error: %v", err)
+			} else {
+				info = fp
+			}
+
+			cancelFn()
+			i.updater(i.info.Name, info)
+			timer.Reset(managerFingerprintInterval)
+		}
+	}
+}
+
+func (i *instanceManager) buildControllerFingerprint(ctx context.Context, base *structs.CSIInfo) (*structs.CSIInfo, error) {
+	if i.client == nil {
+		return nil, fmt.Errorf("No CSI Client")
+	}
+
+	fp := base.Copy()
+
+	healthy, err := i.client.PluginProbe(ctx)
+	if err != nil {
+		return nil, err
+	}
+	fp.SetHealthy(healthy)
+
+	return fp, nil
+}
+
+func (i *instanceManager) buildNodeFingerprint(ctx context.Context, base *structs.CSIInfo) (*structs.CSIInfo, error) {
+	if i.client == nil {
+		return nil, fmt.Errorf("No CSI Client")
```

Comment:
How is this possible? (same q on above method too)

Reply:
If it got called after the instance manager got shut down or if it a test didn't do setup properly. I'll leave a comment saying it's for development rather than anything else